File: Solutions/1942/1942.py
class Solution(object):
    def smallestChair(self, times, targetFriend):
        """
        :type times: List[List[int]]
        :type targetFriend: int
        :rtype: int
        """
        events = []
        AvailChairs = list(range(len(times)))
        OccChairs = []
    
        for i in range(len(times)):
            events.append([times[i][0],i])
            events.append([times[i][1],~i])
    
        events.sort()
    
        for event in events:
            time,friend = event
    
            while (OccChairs and time >= OccChairs[0][0]):
                _,chair = heapq.heappop(OccChairs)
                heapq.heappush(AvailChairs,chair)
            
            if friend >= 0:
                chair = heapq.heappop(AvailChairs)
                if friend == targetFriend:
                    return chair
                heapq.heappush(OccChairs,[times[friend][1], chair])
        return -1

File: Solutions/1942/1942.java
class Solution {
    public int smallestChair(int[][] times, int targetFriend) {
        List<int[]> events = new ArrayList<>();
        PriorityQueue<Integer> availChairs = new PriorityQueue<>();
        PriorityQueue<int[]> occChairs = new PriorityQueue<>((a, b) -> a[0] - b[0]);

        for (int i = 0; i < times.length; i++) {
            events.add(new int[]{times[i][0], i});
            events.add(new int[]{times[i][1], ~i});
        }

        Collections.sort(events, (a, b) -> a[0] - b[0]);

        for (int i = 0; i < times.length; i++) {
            availChairs.add(i);
        }

        for (int[] event : events) {
            int time = event[0];
            int friend = event[1];

            while (!occChairs.isEmpty() && occChairs.peek()[0] <= time) {
                availChairs.add(occChairs.poll()[1]);
            }

            if (friend >= 0) {
                int chair = availChairs.poll();
                if (friend == targetFriend)
                    return chair;
                occChairs.add(new int[]{times[friend][1], chair});
            }
        }
        return -1;
    }
}